### PR TITLE
fix(table): fix scroll delay in header on virtualized table

### PR DESCRIPTION
**Pull request by rivka-ungar**
https://monday.monday.com/boards/3532714909/views/80492480/pulses/7963926689

### CODE REVIEW

File: packages/core/src/components/Table/TableHeader/TableHeader.module.scss [hunk 1]
```diff
@@ -4,7 +4,7 @@
   position: sticky;
   top: 0;
   z-index: 2;
-  background-color: inherit;
+  background-color: var(--primary-background-color);
```
Comment (YossiSaadi) [inline packages/core/src/components/Table/TableHeader/TableHeader.module.scss:7]: need to check how this one behaves with the sticky - on scrolling horizontally and vertically, and hovering a row or a table header cell, doesn't it causing issues?

File: packages/core/src/components/Table/TableVirtualizedBody/TableVirtualizedBody.tsx [hunk 2]
```diff
@@ -95,12 +120,22 @@ const TableVirtualizedBody = forwardRef(
               <List
                 itemSize={RowHeights[size]}
                 height={height}
-                itemCount={items.length}
+                itemCount={virtualizedWithHeader ? items.length + 1 : items.length}
```
Comment (YossiSaadi) [inline packages/core/src/components/Table/TableVirtualizedBody/TableVirtualizedBody.tsx:123]: nice trick

File: packages/core/src/components/Table/TableVirtualizedBody/TableVirtualizedBody.tsx [hunk 3]
```diff
@@ -61,30 +74,42 @@ const TableVirtualizedBody = forwardRef(
...
         });
       },
-      [items, rowRenderer]
+      [items, rowRenderer, virtualizedWithHeader]
     );
 
     useEffect(() => {
-      markTableAsVirtualized();
-    }, [markTableAsVirtualized]);
+      if (!virtualizedWithHeader) markTableAsVirtualized();
```
Comment (YossiSaadi) [inline packages/core/src/components/Table/TableVirtualizedBody/TableVirtualizedBody.tsx:101]: why does it matter though?

File: packages/core/src/components/Table/TableVirtualizedBody/TableVirtualizedBody.tsx [hunk 4]
```diff
@@ -61,30 +74,42 @@ const TableVirtualizedBody = forwardRef(
...
     );
 
     const itemRenderer = useCallback<ComponentType<ListChildComponentProps<TableVirtualizedRow>>>(
       ({ index, style: { width: _width, ...style } }) => {
-        const currentItem = items[index];
+        if (virtualizedWithHeader && index === 0) {
+          return null; //placeholder for virtualized with header
+        }
+        const currentIndex = virtualizedWithHeader ? index - 1 : index;
+        const currentItem = items[currentIndex];
```
Comment (YossiSaadi) [inline packages/core/src/components/Table/TableVirtualizedBody/TableVirtualizedBody.tsx:89]:
```suggestion
        if (virtualizedWithHeader && index === items.length) {
          return null; //placeholder for virtualized with header
        }
        const currentItem = items[index];
```

would this work as well? or do you have to make the header as the index 0? (I know it makes sense it's index 0, but is `innerElementType` just puts it before the first index anyway?)

File: packages/core/src/components/Table/TableVirtualizedBody/TableVirtualizedBody.tsx [hunk 5]
```diff
@@ -95,12 +120,22 @@ const TableVirtualizedBody = forwardRef(
...
+                innerElementType={
+                  virtualizedWithHeader
+                    ? forwardRef(({ children, ...rest }, ref) => (
+                        <div ref={ref} {...rest}>
+                          {headerRenderer(columns)}
+                          {children}
+                        </div>
+                      ))
+                    : undefined
+                }
```
Comment (YossiSaadi) [inline packages/core/src/components/Table/TableVirtualizedBody/TableVirtualizedBody.tsx:138]: is this just lets you wrap everything that VirtualizedList is about to render?
also, do we absolutely need the wrapping div? and what's `rest`? is it coming from VirtualizedList?

Comment (YossiSaadi) [reply packages/core/src/components/Table/TableVirtualizedBody/TableVirtualizedBody.tsx:138]: Can we try Fragments there maybe?
or is it a must? (a yes is a legitimate answer 🤓)

**Review by YossiSaadi — APPROVED** [review]
Thank you.
I know how much frustration it caused and I really appreciate the work you've put into that one 👑

**Comment by YossiSaadi** [comment]
Would also love to try it in Storybook if you have a live example to try locally

File: packages/core/src/components/Table/TableVirtualizedBody/TableVirtualizedBody.tsx (hunk 4 above)
Comment (rivka-ungar) [reply packages/core/src/components/Table/TableVirtualizedBody/TableVirtualizedBody.tsx:89]: `innerElementType` puts it before the first index but if we don't keep an empty slot for it on index 0 it overlaps the first row.

Comment (YossiSaadi) [reply packages/core/src/components/Table/TableVirtualizedBody/TableVirtualizedBody.tsx:89]: got ya

File: packages/core/src/components/Table/TableVirtualizedBody/TableVirtualizedBody.tsx (hunk 3 above)
Comment (rivka-ungar) [reply packages/core/src/components/Table/TableVirtualizedBody/TableVirtualizedBody.tsx:101]: Now that we don't need the sync for the header on the scroll it should act the same as non virtualized table, so there is no need to mark it as virtualized

File: packages/core/src/components/Table/TableHeader/TableHeader.module.scss (hunk 1 above)
Comment (rivka-ungar) [reply packages/core/src/components/Table/TableHeader/TableHeader.module.scss:7]: I checked on all scroll and hover states and it seems to be ok

File: packages/core/src/components/Table/TableVirtualizedBody/TableVirtualizedBody.tsx (hunk 5 above)
Comment (rivka-ungar) [reply packages/core/src/components/Table/TableVirtualizedBody/TableVirtualizedBody.tsx:138]: The div exists in the default behavior of VirtualizedList, so we need to include it here to maintain the current structure. The rest props come from VirtualizedList

**Comment by rivka-ungar** [comment]
> Would also love to try it in Storybook if you have a live example to try locally

I changed the current virtualized story to the new implementation

**Comment by YossiSaadi** [comment]
> > Would also love to try it in Storybook if you have a live example to try locally
> 
> I changed the current virtualized story to the new implementation

Went over this, looks good 💪🏼

**Pull request merged**